In [22]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# 读取数据，并且检查表示是否正确
通常从网络下载的表格，会包含一些错误。所以数据处理的第一步就是确保取得的数据可以处理。这是把数据变成信息的第一步。从TSA（美国运输安全管理局）下载2015，2016年的claims（赔偿申请）的统计表格之后，我们尝试用 **Pandas** 读取它，然后清理数据。需要指出的是，2016年的数据TSA仅提供 PDF 版本的，在从 PDF 转换成 EXCEL 的过程之中，必定会包含大量的错误。我们需要话费大量的时间来处理。

In [23]:
df16 = pd.read_excel('./assets/tsa_claims_data_for_2016_0.xlsx')

In [24]:
df16.head(5)

Claim Number        Date Received   Incident Date Airport Code  \
0            NaN                  NaN             NaN          NaN   
1  2016022629173  2016-02-05 00:00:00  1/2/2016 16:00          ABE   
2  2016052731826  2016-05-18 00:00:00  5/5/2016 17:00          ABE   
3  2016061532244  2016-05-31 00:00:00  5/15/2016 0:00          ABE   
4  2016081633086  2016-07-06 00:00:00  7/5/2016 17:00          ABE   

                                      Airport Name       Airline Name  \
0                      Lehigh Valley International                NaN   
1  Airport, Allentown\nLehigh Valley International  American Airlines   
2  Airport, Allentown\nLehigh Valley International      Allegiant Air   
3  Airport, Allentown\nLehigh Valley International    Delta Air Lines   
4  Airport, Allentown\nLehigh Valley International                UAL   

                Claim Type       Claim Site         Item Category  \
0                      NaN              NaN                   NaN   
1  Passenger Property Loss  Checked Baggage              Currency   
2  Passenger Property Loss  Checked Baggage  Personal Electronics   
3          Property Damage       Checkpoint     Jewelry & Watches   
4          Property Damage  Checked Baggage  Baggage/Cases/Purses   

  Close Amount Disposition Unnamed: 11  
0          NaN         NaN         NaN  
1            0        Deny         NaN  
2            0        Deny         NaN  
3            0        Deny         NaN  
4            -   In Review         NaN

设置一个数字，做一个备份。看看有多少数据被清除掉了。如果清除的数据占比例过大，则需要重新考虑是否能把这个数据转换成有效信息。

In [25]:
rows_before = len(df16) 
print(rows_before)

8688


可以看到最后一列，几乎只包含 **NaN** 的数据，并且没有什么实际意义（估计是用 PDF 转换 Exce l出现的错误列）。需要把它去掉。

In [26]:
df16.drop( columns = 'Unnamed: 11', inplace = True)
df16.head(8)

Claim Number        Date Received   Incident Date Airport Code  \
0            NaN                  NaN             NaN          NaN   
1  2016022629173  2016-02-05 00:00:00  1/2/2016 16:00          ABE   
2  2016052731826  2016-05-18 00:00:00  5/5/2016 17:00          ABE   
3  2016061532244  2016-05-31 00:00:00  5/15/2016 0:00          ABE   
4  2016081633086  2016-07-06 00:00:00  7/5/2016 17:00          ABE   
5  2016100534408  2016-09-22 00:00:00  9/11/2016 0:00          ABE   
6            NaN                  NaN             NaN          NaN   
7  2016081233041  2016-07-06 00:00:00  7/3/2016 19:50          ABQ   

                                      Airport Name       Airline Name  \
0                      Lehigh Valley International                NaN   
1  Airport, Allentown\nLehigh Valley International  American Airlines   
2  Airport, Allentown\nLehigh Valley International      Allegiant Air   
3  Airport, Allentown\nLehigh Valley International    Delta Air Lines   
4  Airport, Allentown\nLehigh Valley International                UAL   
5                               Airport, Allentown    Delta Air Lines   
6                        Albuquerque International                NaN   
7       Sunport Airport\nAlbuquerque International      Allegiant Air   

                Claim Type       Claim Site         Item Category  \
0                      NaN              NaN                   NaN   
1  Passenger Property Loss  Checked Baggage              Currency   
2  Passenger Property Loss  Checked Baggage  Personal Electronics   
3          Property Damage       Checkpoint     Jewelry & Watches   
4          Property Damage  Checked Baggage  Baggage/Cases/Purses   
5          Property Damage  Checked Baggage                 Other   
6                      NaN              NaN                   NaN   
7          Property Damage       Checkpoint  Baggage/Cases/Purses   

  Close Amount    Disposition  
0          NaN            NaN  
1            0           Deny  
2            0           Deny  
3            0           Deny  
4            -      In Review  
5            -      In Review  
6          NaN            NaN  
7            -  *Insufficient

同时看看表格末端，我们发现几行杂乱的数据，需要清除。

In [27]:
df16.tail(10)

Claim Number        Date Received  \
8678                            2017013037046  2016-12-19 00:00:00   
8679                            2017020937401  2017-01-02 00:00:00   
8680                            2016122936166  2016-12-29 00:00:00   
8681  2017020737344\n*Insufficient - Claim is  2016-12-29 00:00:00   
8682                         missing required                  NaN   
8683                    information; claimant                  NaN   
8684                                contacted                  NaN   
8685                      In Review -Claim is                  NaN   
8686                  assigned to an Examiner                  NaN   
8687                    and is being reviewed                  NaN   

        Incident Date Airport Code Airport Name Airline Name  \
8678  11/23/2016 0:00            -            -            -   
8679  11/24/2016 7:22            -            -            -   
8680  12/29/2016 0:00            -            -            -   
8681  12/29/2016 0:00            -            -            -   
8682              NaN          NaN          NaN          NaN   
8683              NaN          NaN          NaN          NaN   
8684              NaN          NaN          NaN          NaN   
8685              NaN          NaN          NaN          NaN   
8686              NaN          NaN          NaN          NaN   
8687              NaN          NaN          NaN          NaN   

                   Claim Type       Claim Site         Item Category  \
8678  Passenger Property Loss  Checked Baggage     Jewelry & Watches   
8679                        -                -    Jewelry & Watches;   
8680                        -                -                     -   
8681  Passenger Property Loss  Checked Baggage  Cosmetics & Grooming   
8682                      NaN              NaN                   NaN   
8683                      NaN              NaN                   NaN   
8684                      NaN              NaN                   NaN   
8685                      NaN              NaN                   NaN   
8686                      NaN              NaN                   NaN   
8687                      NaN              NaN                   NaN   

     Close Amount    Disposition  
8678            -  *Insufficient  
8679            -  *Insufficient  
8680            -  *Insufficient  
8681            -  *Insufficient  
8682          NaN            NaN  
8683          NaN            NaN  
8684          NaN            NaN  
8685          NaN            NaN  
8686          NaN            NaN  
8687          NaN            NaN

In [28]:
df16 = df16.iloc[:-7, :]
df16.tail(5)

Claim Number        Date Received    Incident Date Airport Code  \
8676  2017020737337  2016-12-16 00:00:00  11/19/2016 0:00            -   
8677  2017011936720  2016-11-23 00:00:00  11/23/2016 0:00            -   
8678  2017013037046  2016-12-19 00:00:00  11/23/2016 0:00            -   
8679  2017020937401  2017-01-02 00:00:00  11/24/2016 7:22            -   
8680  2016122936166  2016-12-29 00:00:00  12/29/2016 0:00            -   

     Airport Name Airline Name               Claim Type       Claim Site  \
8676            -            -          Property Damage  Checked Baggage   
8677            -            -  Passenger Property Loss       Checkpoint   
8678            -            -  Passenger Property Loss  Checked Baggage   
8679            -            -                        -                -   
8680            -            -                        -                -   

               Item Category Close Amount    Disposition  
8676  Computer & Accessories            -  *Insufficient  
8677       Jewelry & Watches            -  *Insufficient  
8678       Jewelry & Watches            -  *Insufficient  
8679      Jewelry & Watches;            -  *Insufficient  
8680                       -            -  *Insufficient

在处理的过程中，我们发现有很多行列的数据均，它们的 **Airport Code** 都是 **'-'**，这表示所填写的申请并没有包含机场的信息。我们可以看看这样的数据到底有多少行，若果其占比不大，可以考虑清除。因为我们感兴趣的是关于机场的数据。

In [29]:
no_code = len(df16[ (df16['Airport Code'] == '-')])
print(no_code)

68


在之前定义的 **rows_before**，可以知道总申请大概有8681宗。而从上面的 **no_code** 可以知道那些没有包含机场信息的申请为68宗，所占的比例很小。所以可以出去掉这部分的数据。

In [30]:
df16 = df16[ df16['Airport Code'] != '-' ]

In [31]:
df16.tail(10)

Claim Number        Date Received     Incident Date  Airport Code  \
8603  2016092633849  2016-08-17 00:00:00    7/21/2016 0:00           ZZX   
8604  2016120735929  2016-12-20 00:00:00  11/29/2016 16:00           ZZX   
8605  2017011136464  2016-11-09 00:00:00    9/12/2016 9:15           ZZX   
8606  2016092834058  2016-10-05 00:00:00    1/8/2016 15:02           ZZX   
8607  2016062832462  2016-06-10 00:00:00   1/25/2016 19:20           ZZX   
8608  2016110335478  2016-10-07 00:00:00   6/28/2016 15:46           ZZX   
8609  2016122236143  2016-10-27 00:00:00    8/26/2016 0:00           ZZX   
8610  2016071232610  2016-07-12 00:00:00    5/10/2016 0:00           ZZX   
8618   Claim Number        Date Received     Incident Date  Airport Code   
8671   Claim Number        Date Received     Incident Date  Airport Code   

                          Airport Name  Airline Name       Claim Type  \
8603  vehicle)\nNon TSA Airport (motor             -    Motor Vehicle   
8604  vehicle)\nNon TSA Airport (motor             -    Motor Vehicle   
8605  vehicle)\nNon TSA Airport (motor             -    Motor Vehicle   
8606  vehicle)\nNon TSA Airport (motor             -  Personal Injury   
8607  vehicle)\nNon TSA Airport (motor             -    Motor Vehicle   
8608  vehicle)\nNon TSA Airport (motor             -  Property Damage   
8609  vehicle)\nNon TSA Airport (motor             -    Motor Vehicle   
8610                          vehicle)             -  Personal Injury   
8618                      Airport Name  Airline Name       Claim Type   
8671                      Airport Name  Airline Name       Claim Type   

         Claim Site                   Item Category  Close Amount  \
8603  Motor Vehicle                               -       1348.34   
8604  Motor Vehicle  Automobile Parts & Accessories       1055.95   
8605  Motor Vehicle  Automobile Parts & Accessories             -   
8606  Motor Vehicle  Automobile Parts & Accessories             -   
8607  Motor Vehicle  Automobile Parts & Accessories             -   
8608  Motor Vehicle  Automobile Parts & Accessories             -   
8609  Motor Vehicle  Automobile Parts & Accessories             -   
8610  Motor Vehicle                           Other          1761   
8618     Claim Site                   Item Category  Close Amount   
8671     Claim Site                   Item Category  Close Amount   

          Disposition  
8603  Approve in Full  
8604  Approve in Full  
8605  Closed:Canceled  
8606        In Review  
8607        In Review  
8608        In Review  
8609        In Review  
8610           Settle  
8618      Disposition  
8671      Disposition

在处理数据的过程，一些奇异的行也被发现了。看看上面的 df16.tail() 所显示的最后两列，根本就是 columns 的名称嘛。我们需要一并清理掉。

In [32]:
df16 = df16[ df16['Airport Code'] != 'Airport Code']
df16.tail()

Claim Number        Date Received    Incident Date Airport Code  \
8606  2016092834058  2016-10-05 00:00:00   1/8/2016 15:02          ZZX   
8607  2016062832462  2016-06-10 00:00:00  1/25/2016 19:20          ZZX   
8608  2016110335478  2016-10-07 00:00:00  6/28/2016 15:46          ZZX   
8609  2016122236143  2016-10-27 00:00:00   8/26/2016 0:00          ZZX   
8610  2016071232610  2016-07-12 00:00:00   5/10/2016 0:00          ZZX   

                          Airport Name Airline Name       Claim Type  \
8606  vehicle)\nNon TSA Airport (motor            -  Personal Injury   
8607  vehicle)\nNon TSA Airport (motor            -    Motor Vehicle   
8608  vehicle)\nNon TSA Airport (motor            -  Property Damage   
8609  vehicle)\nNon TSA Airport (motor            -    Motor Vehicle   
8610                          vehicle)            -  Personal Injury   

         Claim Site                   Item Category Close Amount Disposition  
8606  Motor Vehicle  Automobile Parts & Accessories            -   In Review  
8607  Motor Vehicle  Automobile Parts & Accessories            -   In Review  
8608  Motor Vehicle  Automobile Parts & Accessories            -   In Review  
8609  Motor Vehicle  Automobile Parts & Accessories            -   In Review  
8610  Motor Vehicle                           Other         1761      Settle

在完成上面的数据之后，我们继续看完成的表格。结果又发现新的问题，看回 **Airport Name**，结果发现了无法理解的句子，估计是在转换过程中造成的。解决的方法有两个：
1) 逐行校队
2) 因为 Airport Code 没有包含错误，可以考率建立一个 Airport Code 对应 Airport Name 的字典，然后进行转换。
第二个方法显然比较好。可是如果能取得这样的字典呢？也许我们可以从2015年的数据中获得信息（2015年的数据为EXCEL格式，谢谢TSA）。所以我们暂时把它放一边，看看还有什么其他的地方需要清理。

In [33]:
df16['Claim Number'].unique() # 检查 Claim Number 的独立元素

array([nan, 2016022629173, 2016052731826, ..., 2016110335478,
       2016122236143, 2016071232610], dtype=object)

在运行上面的 **command** 之后，发现竟然含有  **NaN**！ **Claim Number** 显示的是每一份赔偿申请的唯一ID，若果连这个信息也没有，估计这个数据在数据库中已经流失了。我们可以检查看看。

In [34]:
df16[ pd.isna(df16['Claim Number']) ]

Claim Number Date Received Incident Date Airport Code  \
0             NaN           NaN           NaN          NaN   
6             NaN           NaN           NaN          NaN   
28            NaN           NaN           NaN          NaN   
37            NaN           NaN           NaN          NaN   
52            NaN           NaN           NaN          NaN   
...           ...           ...           ...          ...   
8281          NaN           NaN           NaN          NaN   
8309          NaN           NaN           NaN          NaN   
8346          NaN           NaN           NaN          NaN   
8437          NaN           NaN           NaN          NaN   
8592          NaN           NaN           NaN          NaN   

                     Airport Name Airline Name Claim Type Claim Site  \
0     Lehigh Valley International          NaN        NaN        NaN   
6       Albuquerque International          NaN        NaN        NaN   
28      Albuquerque International          NaN        NaN        NaN   
37     Southwest Georgia Regional          NaN        NaN        NaN   
52       Alexandria International          NaN        NaN        NaN   
...                           ...          ...        ...        ...   
8281     Sacramento International          NaN        NaN        NaN   
8309     Sacramento International          NaN        NaN        NaN   
8346           Sarasota Bradenton          NaN        NaN        NaN   
8437             Syracuse-Hancock          NaN        NaN        NaN   
8592       Non TSA Airport (motor          NaN        NaN        NaN   

     Item Category Close Amount Disposition  
0              NaN          NaN         NaN  
6              NaN          NaN         NaN  
28             NaN          NaN         NaN  
37             NaN          NaN         NaN  
52             NaN          NaN         NaN  
...            ...          ...         ...  
8281           NaN          NaN         NaN  
8309           NaN          NaN         NaN  
8346           NaN          NaN         NaN  
8437           NaN          NaN         NaN  
8592           NaN          NaN         NaN  

[222 rows x 11 columns]

果然，除了机场名字之外，其他的任何信息都是 **NaN**（缺失或者无法辨识）。这些数据无法提供任何有效的信息，需要清理掉。

In [35]:
df16.dropna(axis = 0, subset = ['Claim Number'], inplace = True)

In [36]:
df16[ pd.isna(df16['Claim Number']) ]

Empty DataFrame
Columns: [Claim Number, Date Received, Incident Date, Airport Code, Airport Name, Airline Name, Claim Type, Claim Site, Item Category, Close Amount, Disposition]
Index: []

在清理之后，我们可以看到已经不包含任何含有 **NaN** 的 **Claim Number** 了。可是这是仍然不够的，在仔细查看 **Airline Name** 之后，我们又发现了转换错误的问题。而且这一次并没有一一对应的字典可以用了（一个机场可以有多个航线开通）。好在我们要分析的是每个机场的赔偿统计，并不涉及航线，所以我们可以简单的把这一列信息去除。

In [37]:
df16.drop( axis = 1, columns = ['Airline Name'], inplace = True)

In [38]:
df16.head()

Claim Number        Date Received   Incident Date Airport Code  \
1  2016022629173  2016-02-05 00:00:00  1/2/2016 16:00          ABE   
2  2016052731826  2016-05-18 00:00:00  5/5/2016 17:00          ABE   
3  2016061532244  2016-05-31 00:00:00  5/15/2016 0:00          ABE   
4  2016081633086  2016-07-06 00:00:00  7/5/2016 17:00          ABE   
5  2016100534408  2016-09-22 00:00:00  9/11/2016 0:00          ABE   

                                      Airport Name               Claim Type  \
1  Airport, Allentown\nLehigh Valley International  Passenger Property Loss   
2  Airport, Allentown\nLehigh Valley International  Passenger Property Loss   
3  Airport, Allentown\nLehigh Valley International          Property Damage   
4  Airport, Allentown\nLehigh Valley International          Property Damage   
5                               Airport, Allentown          Property Damage   

        Claim Site         Item Category Close Amount Disposition  
1  Checked Baggage              Currency            0        Deny  
2  Checked Baggage  Personal Electronics            0        Deny  
3       Checkpoint     Jewelry & Watches            0        Deny  
4  Checked Baggage  Baggage/Cases/Purses            -   In Review  
5  Checked Baggage                 Other            -   In Review

In [39]:
print('The processed data is {0} % of the original data.'.format(100*(len(df16)/rows_before)))

The processed data is 93.45073664825047 % of the original data.


从上图可以看到，除了 **Airport Name**，其他所有的信息都是可分析的，并且删除的数据并不占很大的比例（约为7%）。在接下来，我们读取2015年的数据并清理，然后构造一个机场代码至名字的字典，以帮助我们改进2016年的数据。

In [40]:
df15 = pd.read_excel('./assets/claims-data-2015-as-of-feb-9-2016.xlsx')
df15.head()

Claim Number Date Received Incident D Airport Code  \
0  2015060222904    2015-05-28 2015-05-21          ABE   
1  2015082725115    2015-08-21 2015-08-01          ABE   
2  2015102326219    2015-10-07 2015-10-02          ABE   
3  2015122027695    2015-10-27 2015-10-19          ABE   
4  2015012220065    2015-01-22 2014-12-23          ABI   

                                     Airport Name         Airline Name  \
0  Lehigh Valley International Airport, Allentown        Allegiant Air   
1  Lehigh Valley International Airport, Allentown                    -   
2  Lehigh Valley International Airport, Allentown        Allegiant Air   
3  Lehigh Valley International Airport, Allentown        Allegiant Air   
4                                Abilene Regional  American Airlines     

                Claim Type       Claim Site                  Item Category  \
0          Property Damage       Checkpoint           Personal Electronics   
1          Property Damage  Checked Baggage  Sporting Equipment & Supplies   
2  Passenger Property Loss  Checked Baggage                       Clothing   
3          Property Damage       Checkpoint           Personal Accessories   
4          Property Damage  Checked Baggage                     Home Decor   

  Close Amount      Disposition  
0            0             Deny  
1          180           Settle  
2           60  Approve in Full  
3            -                -  
4            0             Deny

首先需要做的是把 Airport Code 里面为 NaN 的数据全部清理掉。需要注意的是，在2015的数据里面，NAN 一律表示为 **'-'**。

In [41]:
df15 = df15[ df15['Airport Code'] != '-']
df15.head()

Claim Number Date Received Incident D Airport Code  \
0  2015060222904    2015-05-28 2015-05-21          ABE   
1  2015082725115    2015-08-21 2015-08-01          ABE   
2  2015102326219    2015-10-07 2015-10-02          ABE   
3  2015122027695    2015-10-27 2015-10-19          ABE   
4  2015012220065    2015-01-22 2014-12-23          ABI   

                                     Airport Name         Airline Name  \
0  Lehigh Valley International Airport, Allentown        Allegiant Air   
1  Lehigh Valley International Airport, Allentown                    -   
2  Lehigh Valley International Airport, Allentown        Allegiant Air   
3  Lehigh Valley International Airport, Allentown        Allegiant Air   
4                                Abilene Regional  American Airlines     

                Claim Type       Claim Site                  Item Category  \
0          Property Damage       Checkpoint           Personal Electronics   
1          Property Damage  Checked Baggage  Sporting Equipment & Supplies   
2  Passenger Property Loss  Checked Baggage                       Clothing   
3          Property Damage       Checkpoint           Personal Accessories   
4          Property Damage  Checked Baggage                     Home Decor   

  Close Amount      Disposition  
0            0             Deny  
1          180           Settle  
2           60  Approve in Full  
3            -                -  
4            0             Deny

接下来建立需要的字典，用来更正2016年度表格里面的 Airport Name。

In [42]:
dic_aiport = {}
for code in df15['Airport Code'].unique():
    dic_aiport[code] = df15[ df15['Airport Code'] == code]['Airport Name'].unique()[0]

转换 2016 年的数据，利用刚刚取得的字典。

In [43]:
def convert(x):
    # 定义一个方程来转换数据。
    return(dic_aiport.get(x, '-')) # x[3] 是因为 Airport Code 正好在第3列

df16['Airport Name'] = df16['Airport Code'].apply(convert)

In [44]:
df16[df16['Airport Name'] == '-']

Claim Number        Date Received     Incident Date Airport Code  \
38    2016012028439  2016-01-08 00:00:00    1/6/2016 11:10          ABY   
39    2016092834023  2016-09-23 00:00:00    9/13/2016 0:00          ABY   
53    2016022529125  2016-02-03 00:00:00    2/2/2016 10:30          AEX   
109   2017010436286  2016-10-31 00:00:00   10/23/2016 7:00          ART   
613   2016032530114  2016-03-11 00:00:00    3/11/2016 0:00          BHB   
953   2016092133701  2016-08-01 00:00:00    7/17/2016 9:30          BTM   
1154  2016101134585  2016-09-21 00:00:00     9/9/2016 0:00          CDV   
1159  2016040630418  2016-03-23 00:00:00    3/9/2016 10:00          CHO   
1160  2017011936712  2016-11-18 00:00:00   11/11/2016 0:00          CHO   
1191  2016030829426  2016-02-16 00:00:00    1/28/2016 0:00          CIU   
1192  2016031029560  2016-02-18 00:00:00    1/28/2016 0:00          CIU   
1460  2016072032811  2016-06-27 00:00:00    5/14/2016 6:00          COU   
1461  2016052431655  2016-05-10 00:00:00    1/6/2016 12:30          CPR   
1462  2017011336546  2016-11-23 00:00:00   11/22/2016 0:20          CPR   
2199  2016052631762  2016-05-17 00:00:00   1/28/2016 15:00          DVL   
2879  2016092834073  2016-08-10 00:00:00    7/7/2016 12:12          FLO   
2880  2016041930628  2016-03-30 00:00:00   3/15/2016 19:11          FNT   
2881  2016080932978  2016-07-05 00:00:00    6/24/2016 0:00          FNT   
2989  2016040130268  2016-03-18 00:00:00    3/4/2016 12:36          HHH   
2990  2016042930977  2016-04-15 00:00:00     4/3/2016 0:00          HHH   
2991  2016062132344  2016-06-08 00:00:00     6/4/2016 0:00          HHH   
2992  2016052331594  2016-05-09 00:00:00    4/29/2016 0:00          HIB   
3590  2016072832884  2016-06-29 00:00:00   6/18/2016 15:30          INL   
3591  2016110235461  2016-10-08 00:00:00    1/8/2016 11:00          INL   
3592  2016102735254  2016-09-01 00:00:00    8/11/2016 6:00          INL   
3593  2016092133719  2016-09-21 00:00:00    9/21/2016 0:00          INL   
3594  2016051731394  2016-05-17 00:00:00    5/17/2016 0:00          ISN   
3595  2016042930968  2016-08-08 00:00:00     2/4/2016 0:00          ISN   
3609  2016060331991  2016-05-24 00:00:00    5/23/2016 6:30          JAC   
3610  2017011936691  2016-12-01 00:00:00    11/6/2016 6:00          JAC   
4448  2016103135389  2016-09-30 00:00:00    9/25/2016 0:00          LAW   
4449  2016100634494  2016-09-30 00:00:00    9/25/2016 7:00          LAW   
4450  2016101334619  2016-08-29 00:00:00    7/30/2016 0:00          LAW   
5176  2016092233752  2016-07-25 00:00:00    7/21/2016 0:00          MCE   
6129  2016042930971  2016-04-29 00:00:00    4/29/2016 0:00          MRY   
6130  2016100334279  2016-08-16 00:00:00     8/2/2016 0:00          MSL   
6900  2016030329328  2016-02-11 00:00:00   1/15/2016 13:02          PFN   
7268  2016050931177  2016-04-20 00:00:00    4/18/2016 6:00          PIB   
7269  2016091933617  2016-09-19 00:00:00    7/6/2016 14:30          PIB   
7277  2017011736586  2016-11-25 00:00:00   11/22/2016 0:00          PIH   
7342  2016022529135  2016-02-03 00:00:00    1/11/2016 0:00          PRC   
7350  2017020237209  2016-12-24 00:00:00  10/20/2016 10:30          PSM   
7351  2017020837365  2016-12-26 00:00:00  10/27/2016 10:51          PSM   
7369  2016100534393  2016-09-22 00:00:00    9/19/2016 0:00          PUB   
7370  2016031029585  2016-02-16 00:00:00     2/8/2016 0:00          PUW   
7371  2016100634495  2016-08-24 00:00:00    8/23/2016 6:00          PUW   
7388  2016062432397  2016-06-08 00:00:00   5/26/2016 15:30          PVU   
7389  2016101934907  2016-09-08 00:00:00     8/2/2016 0:00          PVU   
7390  2016040630402  2016-03-22 00:00:00   2/11/2016 12:30          PVU   
7391  2016051931516  2016-05-19 00:00:00   4/13/2016 16:00          PVU   
7463  2016052431686  2016-05-13 00:00:00   4/28/2016 17:50          RFD   
7769  2016092133699  2016-07-25 00:00:00     7/1/2016 0:00          SBY   
7770  2016071432673  2016-06-20 

结果我们可以看到，即使应用了字典，仍然有几十家机场的名字信息是 **'-'**，也就是说这些机场很有可能并没有在2015年产生任何的事故和理赔。为了保持数据的最大完整性，我们只有去网上寻找数据，一一的把缺失的名字加进去。

In [45]:
dic_others = {}
codes = df16[df16['Airport Name'] == '-']['Airport Code'].unique()

# code = ['ABY' 'AEX' 'ART' 'BHB' 'BTM' 'CDV' 'CHO' 'CIU' 'COU' 'CPR' 'DVL' 'FLO'
#                'FNT' 'HHH' 'HIB' 'INL' 'ISN' 'JAC' 'LAW' 'MCE' 'MRY' 'MSL' 'PFN' 'PIB'
#               PIH' 'PRC' 'PSM' 'PUB' 'PUW' 'PVU' 'RFD' 'SBY' 'SJT']

names = ['Southwest Georgia Regional Airport', 'Alexandria International Airport', 'Watertown International Airport', \
         'Hancock County-Bar Harbor Airport', 'Bert Mooney Airport', 'Merle K. (Mudhole) Smith Airport', 'Charlottesville–Albemarle Airport', \
         'Chippewa County International Airport', 'Columbia Regional Airport', 'Casper–Natrona County International Airport', \
         'Devils Lake Regional Airport', 'Florence Regional Airport', 'Bishop International Airport', 'Hilton Head Airport', \
         'Range Regional Airport', 'Falls International Airport', 'Sloulin Field International Airport', 'Jackson Hole Airport', \
         'Lawton–Fort Sill Regional Airport', 'Merced Regional Airport', 'Monterey Regional Airport', 'Northwest Alabama Regional Airport', \
         'Panama City–Bay County International Airport', 'Hattiesburg–Laurel Regional Airport', 'Pocatello Regional Airport', \
         'Ernest A. Love Field', 'Portsmouth International Airport at Pease', 'Pueblo Memorial Airport', 'Pullman–Moscow Regional Airport', \
         'Provo Municipal Airport', 'Chicago Rockford International Airport', 'Salisbury–Ocean City–Wicomico Regional Airport', \
         'San Angelo Regional Airport']

dic_others = dict( zip(codes, names))

for c in codes:
    df16.loc[ df16['Airport Code'] == c, 'Airport Name'] =   df16[ df16['Airport Code']  == c ]['Airport Code'].apply( lambda x: dic_others[x] )

In [46]:
df16[df16['Airport Name'] == '-']

Empty DataFrame
Columns: [Claim Number, Date Received, Incident Date, Airport Code, Airport Name, Claim Type, Claim Site, Item Category, Close Amount, Disposition]
Index: []

我们可以看到，在完成这个转换之后，机场名称这一列已经没有确实信息了。接下来我们就要关注最主要的信息了，赔偿申请处理结果( **Disposition** )。在比较两个表格之后（下列所示），我们发现两个表格在 Disposition 这一栏所含有的独立变量的数量是不同的。原因是2016年把处理结果更加的细化了。为了更好的处理数据，需要两个表格的在这一栏的变量数目一致，而我们无法对2015的数据进行任何细化（无法获得原始数据），所以我们只能对2016的数据进行修改。

In [47]:
print('Disposition types in 2015: {0}, {1}, {2}, {3}'.format( *list(df15['Disposition'].unique()) ) )

Disposition types in 2015: Deny, Settle, Approve in Full, -


In [48]:
print('Disposition types in 2016: {0}, {1}, {2}, {3}, {4}, {5}, {6}'.format( *list(df16['Disposition'].unique()) ) )

Disposition types in 2016: Deny, In Review, *Insufficient, Approve in Full, Settle, Closed:Canceled, Closed:Contractor Claim


In [66]:
df16.loc[ (df16['Disposition'] != 'Deny') & (df16['Disposition'] != 'Approve in Full') & (df16['Disposition'] != 'Settle'), 'Disposition'] = '-'
df16.head(6)

Claim Number        Date Received   Incident Date Airport Code  \
1  2016022629173  2016-02-05 00:00:00  1/2/2016 16:00          ABE   
2  2016052731826  2016-05-18 00:00:00  5/5/2016 17:00          ABE   
3  2016061532244  2016-05-31 00:00:00  5/15/2016 0:00          ABE   
4  2016081633086  2016-07-06 00:00:00  7/5/2016 17:00          ABE   
5  2016100534408  2016-09-22 00:00:00  9/11/2016 0:00          ABE   
7  2016081233041  2016-07-06 00:00:00  7/3/2016 19:50          ABQ   

                                     Airport Name               Claim Type  \
1  Lehigh Valley International Airport, Allentown  Passenger Property Loss   
2  Lehigh Valley International Airport, Allentown  Passenger Property Loss   
3  Lehigh Valley International Airport, Allentown          Property Damage   
4  Lehigh Valley International Airport, Allentown          Property Damage   
5  Lehigh Valley International Airport, Allentown          Property Damage   
7       Albuquerque International Sunport Airport          Property Damage   

        Claim Site         Item Category Close Amount Disposition  
1  Checked Baggage              Currency            0        Deny  
2  Checked Baggage  Personal Electronics            0        Deny  
3       Checkpoint     Jewelry & Watches            0        Deny  
4  Checked Baggage  Baggage/Cases/Purses            -           -  
5  Checked Baggage                 Other            -           -  
7       Checkpoint  Baggage/Cases/Purses            -           -

最后，像2016的表格那样，我们也把2015年的航空公司名称去除。

In [67]:
df15.drop( columns = ['Airline Name'], axis = 1, inplace = True)

In [68]:
df15.head(5)

Claim Number Date Received Incident D Airport Code  \
0  2015060222904    2015-05-28 2015-05-21          ABE   
1  2015082725115    2015-08-21 2015-08-01          ABE   
2  2015102326219    2015-10-07 2015-10-02          ABE   
3  2015122027695    2015-10-27 2015-10-19          ABE   
4  2015012220065    2015-01-22 2014-12-23          ABI   

                                     Airport Name               Claim Type  \
0  Lehigh Valley International Airport, Allentown          Property Damage   
1  Lehigh Valley International Airport, Allentown          Property Damage   
2  Lehigh Valley International Airport, Allentown  Passenger Property Loss   
3  Lehigh Valley International Airport, Allentown          Property Damage   
4                                Abilene Regional          Property Damage   

        Claim Site                  Item Category Close Amount  \
0       Checkpoint           Personal Electronics            0   
1  Checked Baggage  Sporting Equipment & Supplies          180   
2  Checked Baggage                       Clothing           60   
3       Checkpoint           Personal Accessories            -   
4  Checked Baggage                     Home Decor            0   

       Disposition  
0             Deny  
1           Settle  
2  Approve in Full  
3                -  
4             Deny

In [69]:
df15.to_csv('claimns_2015.csv')
df16.to_csv('claimns_2016.csv')